File: backend/app/schemas/validation.py
"""Validation evidence, separate from input structure and solver result tables."""
from typing import Literal

from .base import PayloadModel


class ValidationHelp(PayloadModel):
    rule: str
    steps: list[str]


class ValidationIssue(PayloadModel):
    code: str
    section: str
    severity: Literal['error', 'warning']
    message: str
    table: str | None = None
    row: list[str] | None = None
    period: str | None = None
    area: str | None = None
    help: ValidationHelp | None = None
    actual: object = None
    expected: object = None


class ValidationSection(PayloadModel):
    id: str
    title: str
    error_count: int
    warning_count: int
    fillable_count: int | None = None


class ScenarioValidation(PayloadModel):
    revision: str | None = None
    state: str | None = None
    valid: bool | None = None
    issues: list[ValidationIssue] | None = None
    error_count: int | None = None
    warning_count: int | None = None
    truncated: bool | None = None
    catalog_version: int | None = None
    model_version: str | None = None
    model_check: str | None = None
    feasibility: str | None = None
    termination_condition: str | None = None
    slacks_disabled: bool | None = None
    sections: list[ValidationSection] | None = None
    units: dict[str, str] | None = None
    periods: list[str] | None = None
    missing_tables: list[str] | None = None
    tables_with_issues: list[str] | None = None
    check_for_missing_tables: bool | None = None
    check_for_minimum_required_tables: bool | None = None
    check_for_infeasibility: bool | None = None
    error: str | None = None


class ConstraintViolationRecord(PayloadModel):
    violation: int | float
    side: str | None
    constraint: str
    lower_bound: int | float | None
    body_value: int | float | None
    upper_bound: int | float | None


class ConstraintViolationsSummary(PayloadModel):
    count: int
    tolerance: int | float
    relative_tolerance: int | float | None = None
    relative_tolerance_scope: Literal['linear_currency_equalities'] | None = None
    violations: list[ConstraintViolationRecord]
    logged_count: int | None = None
    status: Literal['complete', 'partial', 'unavailable'] | None = None
    evaluated_count: int | None = None
    skipped_count: int | None = None
    returned_count: int | None = None
    truncated: bool | None = None
    reason: str | None = None
    solution_state: Literal['solver_solution', 'current_model_values'] | None = None

File: electron/ui/src/types/validation.ts
export interface ConstraintViolationRecord {
  violation: number;
  side: string | null;
  constraint: string;
  lower_bound: number | null;
  body_value: number | null;
  upper_bound: number | null;
}

export interface ConstraintViolationsSummary {
  count: number;
  tolerance: number;
  relative_tolerance?: number;
  relative_tolerance_scope?: 'linear_currency_equalities';
  logged_count?: number;
  status?: "complete" | "partial" | "unavailable";
  evaluated_count?: number;
  skipped_count?: number;
  returned_count?: number;
  truncated?: boolean;
  reason?: string;
  solution_state?: "solver_solution" | "current_model_values";
  violations: ConstraintViolationRecord[];
}

export interface ValidationSection {
  id: string;
  title: string;
  error_count: number;
  warning_count: number;
  fillable_count?: number;
}

/** Completeness, model construction and feasibility are independent evidence. */
export interface ScenarioValidation {
  revision?: string;
  state?: 'not_checked' | 'inputs_complete' | 'needs_input' | 'outdated' | 'model_built'
    | 'not_determined' | 'infeasible' | 'feasible' | 'build_failed';
  catalog_version?: number;
  model_version?: string;
  termination_condition?: string;
  slacks_disabled?: boolean;
  issues?: ValidationIssue[];
  error_count?: number;
  warning_count?: number;
  truncated?: boolean;
  model_check?: 'not_run' | 'passed' | 'failed';
  feasibility?: 'not_run' | 'not_supported' | 'not_determined' | 'infeasible' | 'feasible';
  sections?: ValidationSection[];
  units?: Record<string, string>;
  periods?: string[];
  valid?: boolean | null;
  missing_tables?: string[];
  tables_with_issues?: string[];
  check_for_missing_tables?: boolean | null;
  check_for_minimum_required_tables?: boolean | null;
  check_for_infeasibility?: boolean | null;
  error?: string | null;
}

export interface ValidationIssue {
  code: string;
  section: string;
  severity: 'error' | 'warning';
  message: string;
  table?: string | null;
  row?: string[];
  period?: string | null;
  area?: string;
  actual?: unknown;
  expected?: unknown;
  help?: {rule: string; steps: string[]} | null;
}

export interface ScenarioFillPreview {
  revision: string;
  value: number;
  cell_count: number;
  tables: Array<{name: string; cell_count: number; unit: string}>;
}
